core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java

```diff
@@ -849,4 +850,33 @@ private String toStringSlow(@Nullable Channel ch, @Nullable RequestLogAccess par
             return buf.toString();
         }
     }
+
+    @Override
+    public CompletableFuture<Void> initiateConnectionShutdown() {
+        final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+
+        setAdditionalRequestHeader(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+        log().whenRequestComplete().handle((ignore, ex) -> {
+            if (ex != null) {
+                completableFuture.completeExceptionally(ex);
+                return null;
+            }
+            final Channel ch = channel();
+            if (ch == null) {
+                completableFuture.complete(null);
+                return null;
+            }
+
+            ch.closeFuture().addListener(f -> {
+                if (f.cause() == null) {
+                    completableFuture.complete(null);
+                } else {
+                    completableFuture.completeExceptionally(f.cause());
+                }
+            });
+            HttpSession.get(ch).deactivate();
```

Comment:
If I understand correctly this isn't needed anymore. Can you confirm @ikhoon ? 🙇 

```suggestion
```

Reply:
>  we want the next request to trigger shutdown

Exactly. That was the intention.

Reply:
How about adding a comment then?
`// To deactivate the channel when initiateShutdown is called after the RequestHeaders is sent. The next request will trigger shutdown.`